src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationOptions.cs

```diff
@@ -0,0 +1,169 @@
+        /// </summary>
+        public SuperOfficeAuthenticationEnvironment Environment
+        {
+            get
+            {
+                return this._environment;
+            }
+
+            set
+            {
+                _environment = value;
+                UpdateEndpoints();
+            }
+        }
+
+        /// <summary>
+        /// When true user claims will include SuperOffice function rights.
+        /// </summary>
+        public bool IncludeFunctionalRightsAsClaims { get; set; }
+
+        /// <summary>
+        /// When true user claims will include SuperOffice id_token claims.
+        /// </summary>
+        public bool IncludeIdTokenAsClaims { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether to validate tokens using SuperOffice's public key.
+        /// </summary>
+        public bool ValidateTokens { get; set; } = true;
+
+        /// <summary>
+        /// Gets or sets the default period of time to cache the SuperOffice public key(s)
+        /// retrieved from the endpoint specified by <see cref="JwksEndpoint"/>.
+        /// </summary>
+        /// <remarks>
+        /// The default public key cache lifetime is 15 minutes.
+        /// </remarks>
+        public TimeSpan JwksCacheLifetime { get; set; } = TimeSpan.FromMinutes(15);
+
+        /// <inheritdoc />
+        public override void Validate()
+        {
+            base.Validate();
+
+            if (_environment == SuperOfficeAuthenticationEnvironment.Production
+                && !AuthorizationEndpoint.StartsWith("https://", System.StringComparison.InvariantCultureIgnoreCase))
+            {
+                throw new NotSupportedException("Production environment requires secure endpoints, i.e. begins with 'https://'.");
+            }
+        }
+
+        /// <summary>
+        /// Updates configuration settings.
+        /// </summary>
+        private void UpdateEndpoints()
+        {
+            var env = GetEnvironment();
+
+            AuthorizationEndpoint = string.Format(CultureInfo.InvariantCulture,
+                SuperOfficeAuthenticationConstants.FormatStrings.AuthorizeEndpoint,
+                env);
+
+            TokenEndpoint = string.Format(CultureInfo.InvariantCulture,
+                SuperOfficeAuthenticationConstants.FormatStrings.TokenEndpoint,
+                env);
+
+            ClaimsIssuer = string.Format(CultureInfo.InvariantCulture,
+                SuperOfficeAuthenticationConstants.FormatStrings.ClaimsIssuer,
+                env);
+
+            // UserInformationEndpoint will include context identifier after authentication in SuperOfficeAuthenticationHandler.CreateTicketAsync
+            UserInformationEndpoint = string.Concat(
+                string.Format(CultureInfo.InvariantCulture, SuperOfficeAuthenticationConstants.FormatStrings.ClaimsIssuer, env),
+                SuperOfficeAuthenticationConstants.FormatStrings.UserInfoEndpoint);
```

Comment:
This would be more robust if built-up using an `UriBuilder` to combine the host and path, something like this:

```
string host = string.Format(
    CultureInfo.InvariantCulture,
    SuperOfficeAuthenticationConstants.FormatStrings.ClaimsIssuer,
    env);
var builder = new UriBuilder(host)
{
    Path = SuperOfficeAuthenticationConstants.FormatStrings.UserInfoEndpoint,
};

UserInformationEndpoint = builder.ToString();
};
```

Reply:
Yeah, that's a big heavy for such a small thing. This code has been simplified.